File: minigit-3/py/commands/add.py
"""add command - Stage files for commit."""

import sys
import os
from pathlib import Path


def run(args: list[str]) -> int:
    """Stage files for commit."""
    from utils import find_repo_root
    from objects import create_blob
    from index import read_index, write_index, add_to_index, remove_from_index
    from refs import get_head_sha

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse flags
    all_flag = False
    update_flag = False
    paths = []

    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ('-A', '--all'):
            all_flag = True
        elif arg in ('-u', '--update'):
            update_flag = True
        else:
            paths.append(arg)
        i += 1

    if not paths and not all_flag and not update_flag:
        print("error: nothing specified, nothing added", file=sys.stderr)
        return 1

    # Get currently tracked files (from index)
    entries = read_index(repo_root)
    tracked_files = {e['path'] for e in entries}

    # Get files from last commit
    committed_files = set()
    head_sha = get_head_sha(repo_root)
    if head_sha:
        from objects import read_commit, read_tree

        def get_tree_files(tree_sha: str, prefix: str = '') -> set:
            files = set()
            try:
                tree_entries = read_tree(repo_root, tree_sha)
                for entry in tree_entries:
                    path = f"{prefix}{entry.name}" if prefix else entry.name
                    if entry.mode == '40000':
                        files.update(get_tree_files(entry.sha, path + '/'))
                    else:
                        files.add(path)
            except:
                pass
            return files

        try:
            commit = read_commit(repo_root, head_sha)
            committed_files = get_tree_files(commit['tree'])
        except:
            pass

    all_tracked = tracked_files | committed_files

    # Collect files to add
    files_to_add = []

    if all_flag or update_flag:
        # Walk working directory
        for root, dirs, files in os.walk(repo_root):
            # Skip .minigit directory
            if '.minigit' in dirs:
                dirs.remove('.minigit')

            for file in files:
                abs_path = Path(root) / file
                rel_path = str(abs_path.relative_to(repo_root))

                if all_flag:
                    files_to_add.append(rel_path)
                elif update_flag:
                    if rel_path in all_tracked:
                        files_to_add.append(rel_path)

        # Handle deletions
        if all_flag or update_flag:
            for tracked in all_tracked:
                full_path = repo_root / tracked
                if not full_path.exists():
                    # File was deleted
                    remove_from_index(repo_root, tracked)

    # Add explicit paths
    for path_arg in paths:
        path = Path(path_arg)

        # Handle absolute paths
        if path.is_absolute():
            try:
                path = path.relative_to(repo_root)
            except ValueError:
                print(f"error: '{path_arg}' is outside repository", file=sys.stderr)
                return 1

        full_path = repo_root / path

        if not full_path.exists():
            # Check if it's a tracked file being deleted
            rel_str = str(path)
            if rel_str in all_tracked:
                remove_from_index(repo_root, rel_str)
                continue

            print(f"error: pathspec '{path_arg}' did not match any files", file=sys.stderr)
            return 1

        if full_path.is_file():
            files_to_add.append(str(path))
        elif full_path.is_dir():
            # Add all files in directory
            for root, dirs, files in os.walk(full_path):
                if '.minigit' in dirs:
                    dirs.remove('.minigit')
                for file in files:
                    abs_file = Path(root) / file
                    rel_file = str(abs_file.relative_to(repo_root))
                    files_to_add.append(rel_file)

    # Add files to index
    for rel_path in files_to_add:
        full_path = repo_root / rel_path

        if not full_path.exists():
            continue

        if full_path.is_symlink():
            # Handle symlink
            target = os.readlink(full_path)
            content = target.encode()
            mode = '120000'
        else:
            content = full_path.read_bytes()
            # Check if executable
            stat = full_path.stat()
            if stat.st_mode & 0o111:
                mode = '100755'
            else:
                mode = '100644'

        sha = create_blob(repo_root, content)
        add_to_index(repo_root, rel_path, sha, mode, stat if not full_path.is_symlink() else None)

    return 0

File: minigit-3/ts/src/commands/add.ts
// add command - Stage files for commit

import * as fs from 'fs';
import * as path from 'path';
import { IndexEntry } from '../types';
import { findRepoRoot, getFileMode, normalizePath } from '../utils';
import { writeBlob } from '../objects';
import { readIndex, writeIndex, removeFromIndex } from '../index-file';

export function add(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  let pathspecs: string[] = [];
  let updateOnly = false;
  let all = false;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-A' || arg === '--all') {
      all = true;
    } else if (arg === '-u' || arg === '--update') {
      updateOnly = true;
    } else {
      pathspecs.push(arg);
    }
  }

  if (all) {
    pathspecs = ['.'];
  }

  if (pathspecs.length === 0 && !updateOnly) {
    console.error('Nothing specified, nothing added.');
    process.exit(0);
  }

  const entries = readIndex(repoRoot);
  const trackedFiles = new Set(entries.map(e => e.name));
  const newEntries = new Map<string, IndexEntry>();

  // Copy existing entries
  for (const entry of entries) {
    newEntries.set(entry.name, entry);
  }

  // Get all files to potentially add
  const filesToProcess: string[] = [];

  if (updateOnly) {
    // Only process tracked files
    for (const entry of entries) {
      const fullPath = path.join(repoRoot, entry.name);
      if (fs.existsSync(fullPath)) {
        filesToProcess.push(entry.name);
      } else {
        // File deleted - mark for removal
        newEntries.delete(entry.name);
      }
    }
  } else {
    // Process specified pathspecs
    for (const pathspec of pathspecs) {
      const absPath = path.resolve(process.cwd(), pathspec);
      const relPath = path.relative(repoRoot, absPath);

      if (!fs.existsSync(absPath)) {
        console.error(`fatal: pathspec '${pathspec}' did not match any files`);
        process.exit(1);
      }

      if (fs.statSync(absPath).isDirectory()) {
        // Add all files in directory
        addDirectory(absPath, repoRoot, filesToProcess, all);
      } else {
        filesToProcess.push(normalizePath(relPath));
      }
    }
  }

  // Process each file
  for (const relPath of filesToProcess) {
    const absPath = path.join(repoRoot, relPath);

    if (!fs.existsSync(absPath)) {
      // File deleted
      if (all || updateOnly) {
        newEntries.delete(relPath);
      }
      continue;
    }

    const stats = fs.statSync(absPath);

    if (stats.isSymbolicLink()) {
      // Handle symlink
      const target = fs.readlinkSync(absPath);
      const sha = writeBlob(Buffer.from(target), repoRoot);
      const entry = createIndexEntry(relPath, sha, 0o120000, stats);
      newEntries.set(relPath, entry);
    } else if (stats.isFile()) {
      const content = fs.readFileSync(absPath);
      const sha = writeBlob(content, repoRoot);
      const mode = getFileMode(absPath);
      const entry = createIndexEntry(relPath, sha, mode, stats);
      newEntries.set(relPath, entry);
    }
  }

  // Handle -A flag: also remove deleted files
  if (all) {
    for (const entry of entries) {
      const fullPath = path.join(repoRoot, entry.name);
      if (!fs.existsSync(fullPath)) {
        newEntries.delete(entry.name);
      }
    }
  }

  // Write updated index
  writeIndex(Array.from(newEntries.values()), repoRoot);
}

function addDirectory(dirPath: string, repoRoot: string, files: string[], includeNew: boolean): void {
  const entries = fs.readdirSync(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    // Skip .minigit directory
    if (entry.name === '.minigit') continue;

    const fullPath = path.join(dirPath, entry.name);
    const relPath = normalizePath(path.relative(repoRoot, fullPath));

    if (entry.isDirectory()) {
      addDirectory(fullPath, repoRoot, files, includeNew);
    } else if (entry.isFile() || entry.isSymbolicLink()) {
      files.push(relPath);
    }
  }
}

function createIndexEntry(name: string, sha: string, mode: number, stats: fs.Stats): IndexEntry {
  const now = Math.floor(Date.now() / 1000);
  return {
    ctimeSec: Math.floor(stats.ctimeMs / 1000),
    ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
    mtimeSec: Math.floor(stats.mtimeMs / 1000),
    mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
    dev: stats.dev,
    ino: stats.ino,
    mode,
    uid: stats.uid,
    gid: stats.gid,
    size: stats.size,
    sha,
    flags: Math.min(name.length, 0xfff),
    name
  };
}
